activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/agenda/TakeOutgoingSequenceFlowsOperation.java

```diff
@@ -299,7 +300,15 @@ protected void cleanupCompensation() {
     }
 
     protected void cleanupExecutions(FlowElement currentFlowElement) {
-        if (execution.getParentId() != null && execution.isScope()) {
+        if (execution.getParentId() != null && execution.isScope() && execution.hasInstances()) {
+            for (TaskEntity taskEntity : commandContext.getTaskEntityManager().findTasksByProcessInstanceId(execution.getProcessInstanceId()))
```

Comment:
what about other types of tasks such as manual task, integration tasks, etc! Shouldn't that apply to them?

Reply:
I believe that this problem will only happen with UserTask types because it origins in the [insert(taskEntity,executionEntity) of the class TaskEntityManagerImpl.java](https://github.com/Activiti/Activiti/blob/eeda8439bcd30f525f10b23f54a7cbb4819d2ef9/activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/persistence/entity/TaskEntityManagerImpl.java#L75) that is only being called in the [UserTaskActivityBehavior](https://github.com/Activiti/Activiti/blob/096fb2ceefa005dddf7b9a613dd0e0b1416d2981/activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/behavior/UserTaskActivityBehavior.java#L212). There's another reference to this method in the [SaveTaskCmd.java](https://github.com/Activiti/Activiti/blob/407f917f216c03580cf99757c5eeab0d946bf5f2/activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/cmd/SaveTaskCmd.java#L54) but since the `executionEntity` is being passed as `null`, it's not really a problem. The issue happened when adding an execution id to a `taskEntity` from an execution that was already deleted.

Reply:
I removed the above changes to debug the test and better understand the issue. However, the test was green. The test doesn't seem to replicate the issue.

Additionally, can you remove any unrelated data/steps to the issue from the process definition? That can help us better understand debug and understand the issue.

Reply:
The `for` loop block doesn't seem to get executed. When the `currentDistributionList` list is empty the `else` block is always executed.

Also, I noticed that process has 2 sequences leading to the subprocess (2 parallel gateways after Distribution Task). Which could cause issues.

Reply:
Seems that the problem is due the fact user task insert is executed after the process `execution` is being deleted. I'm not sure why the insert is happening in the first place.

Reply:
This was working in APS 1.11.4 that is using Activiti 5.23.0.1 and stopped working in APS 2.3.4 that is using Activiti 7.4.0. 
If the `nrOfInstances` is 0 (that only happens if the `currentDistributionList` is empty) in the MultiInstanceActivityBehavior.java it leaves the execution without executing anything. Therefore I added the `execution.hasInstances()` in this method `cleanupExecutions()` to avoid the execution record to be deleted from the database and to be linked with the task. That's what was being done in the previous version of the Activiti where the case was working. 
The problem was that the execution was being created in memory, was being associated with the task and then the execution was being deleted causing a database error when storing the task record in the database (because the execution was not created in the database).
To fix this, I compared both Activiti versions and after the fix I tested the process to check if workflow was the same between the APS 1.11.4 and APS 2.3.4.